builder/amazon/common/ssm_driver.go

```diff
@@ -48,18 +50,33 @@ func NewSSMDriver(config SSMDriverConfig) *SSMDriver {
 func (d *SSMDriver) StartSession(ctx context.Context, input ssm.StartSessionInput) (*ssm.StartSessionOutput, error) {
 	log.Printf("Starting PortForwarding session to instance %q", aws.StringValue(input.Target))
 
-	var output *ssm.StartSessionOutput
-	err := retry.Config{
-		ShouldRetry: func(err error) bool { return IsAWSErr(err, "TargetNotConnected", "") },
-		RetryDelay:  (&retry.Backoff{InitialBackoff: 200 * time.Millisecond, MaxBackoff: 60 * time.Second, Multiplier: 2}).Linear,
-	}.Run(ctx, func(ctx context.Context) (err error) {
-		output, err = d.SvcClient.StartSessionWithContext(ctx, &input)
-		return err
-	})
+	output, err := d.StartSessionWithContext(ctx, input)
 	if err != nil {
 		return nil, fmt.Errorf("error encountered in starting session for instance %q: %s", aws.StringValue(input.Target), err)
 	}
 
+	d.retryAfterTermination = make(chan bool, 1)
+	// Starts go routine that will keep listening to channels and retry the session creation/connection when needed.
+	// The log polling process will add data to the channels whenever a retryable error happens to the session or if it's terminated.
+	go func(ctx context.Context, driver *SSMDriver, input ssm.StartSessionInput) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case r, ok := <-driver.retryAfterTermination:
+				if !ok {
+					return
+				}
+				if r {
+					log.Printf("[DEBUG] Attempting to restablishing SSM connection")
+					_, _ = driver.StartSession(ctx, input)
```

Comment:
Ah, this is calling, `d.StartSession`; this is like a nested goroutine recursive call on a func that changes fields of `d`. This is going to cause a loooooot of race conditions, hence why Moss was adding the waitgroup there. Testing this with a `-race` option should scream pretty loud.

Reply:
Just to close the loop here @azr. The race detector currently doesn't report any issues. But I can see the concern with the recursive call modifying state. Lets see if/how we can refactor during our pairing session.